jetty-core/jetty-siwe/src/main/java/org/eclipse/jetty/security/siwe/EthereumAuthenticator.java

```diff
@@ -0,0 +1,772 @@
+    {
+        // if this is a request resulting from a redirect after auth is complete
+        // (ie its from a redirect to the original request uri) then due to
+        // browser handling of 302 redirects, the method may not be the same as
+        // that of the original request. Replace the method and original post
+        // params (if it was a post).
+        if (authenticationState instanceof AuthenticationState.Succeeded)
+        {
+            Session session = request.getSession(false);
+            if (session == null)
+                return request; //not authenticated yet
+
+            // Remove the nonce set used for authentication.
+            session.removeAttribute(NONCE_SET_ATTR);
+
+            HttpURI juri = (HttpURI)session.getAttribute(J_URI);
+            HttpURI uri = request.getHttpURI();
+            if ((uri.equals(juri)))
+            {
+                session.removeAttribute(J_URI);
+
+                Fields fields = (Fields)session.removeAttribute(J_POST);
+                if (fields != null)
+                    request.setAttribute(FormFields.class.getName(), fields);
+
+                String method = (String)session.removeAttribute(J_METHOD);
+                if (method != null && request.getMethod().equals(method))
+                {
+                    return new Request.Wrapper(request)
+                    {
+                        @Override
+                        public String getMethod()
+                        {
+                            return method;
+                        }
+                    };
+                }
+            }
+        }
+
+        return request;
+    }
+
+    @Override
+    public Constraint.Authorization getConstraintAuthentication(String pathInContext, Constraint.Authorization existing, Function<Boolean, Session> getSession)
+    {
+        if (isAuthenticationRequest(pathInContext))
+            return Constraint.Authorization.ANY_USER;
+        if (isLoginPage(pathInContext) || isErrorPage(pathInContext))
+            return Constraint.Authorization.ALLOWED;
+        return existing;
+    }
+
+    protected String readMessage(InputStream in) throws IOException
+    {
+        Iso88591StringBuilder out = new Iso88591StringBuilder();
+        byte[] buffer = new byte[1024];
+        int totalRead = 0;
+
+        while (true)
+        {
+            int len = in.read(buffer, 0, buffer.length);
+            if (len < 0)
+                break;
+
+            totalRead += len;
+            if (totalRead > _maxMessageSize)
+                throw new BadMessageException("SIWE Message Too Large");
+            out.append(buffer, 0, len);
+        }
+
+        return out.build();
+    }
+
+    protected SignedMessage parseMessage(Request request, Response response, Callback callback)
+    {
+        try
+        {
+            InputStream inputStream = Content.Source.asInputStream(request);
```

Comment:
Why going blocking here?

Can core authentication be asynchronous?

Reply:
No way to do async authentication currently.

Reply:
@lachlan-roberts project for you :)